src/sources/util/http_scrape.rs

```diff
@@ -0,0 +1,199 @@
+///   - The HTTP response is decoded/parsed into events by the specific context.
+///   - The events are then sent to the output stream.
+pub(crate) async fn http_scrape<H: HttpScraper + std::marker::Send + Clone>(
+    inputs: GenericHttpScrapeInputs,
+    context: H,
+    mut out: SourceSender,
+) -> Result<(), ()> {
+    let mut stream = IntervalStream::new(tokio::time::interval(Duration::from_secs(
+        inputs.interval_secs,
+    )))
+    .take_until(inputs.shutdown)
+    .map(move |_| stream::iter(inputs.urls.clone()))
+    .flatten()
+    .map(move |url| {
+        // Building the HttpClient should not fail as it is just setting up the client with the
+        // proxy and tls settings.
+        let client = HttpClient::new(inputs.tls.clone(), &inputs.proxy)
+            .expect("Building HTTP client failed");
+        let endpoint = url.to_string();
+
+        let mut context = context.clone();
+        context.build(&url);
+
+        let mut builder = Request::get(&url).header(http::header::ACCEPT, &inputs.content_type);
+
+        // add user supplied headers
+        for (header, values) in &inputs.headers {
+            for value in values {
+                builder = builder.header(header, value);
+            }
+        }
+
+        // building an empty request should be infallible
+        let mut request = builder.body(Body::empty()).expect("error creating request");
+
+        if let Some(auth) = &inputs.auth {
+            auth.apply(&mut request);
+        }
+
+        let start = Instant::now();
+        client
+            .send(request)
+            .map_err(Error::from)
+            .and_then(|response| async move {
+                let (header, body) = response.into_parts();
+                let body = hyper::body::to_bytes(body).await?;
+                emit!(EndpointBytesReceived {
+                    byte_size: body.len(),
+                    protocol: "http",
+                    endpoint: endpoint.as_str(),
+                });
+                Ok((header, body))
+            })
+            .into_stream()
+            .filter_map(move |response| {
+                ready(match response {
+                    Ok((header, body)) if header.status == hyper::StatusCode::OK => {
+                        emit!(RequestCompleted {
+                            start,
+                            end: Instant::now()
+                        });
+                        match context.on_response(&url, &header, &body) {
+                            Some(events) => {
+                                emit!(HttpScrapeEventsReceived {
+                                    byte_size: events.size_of(),
+                                    count: events.len(),
+                                    url: url.to_string()
+                                });
+                                Some(stream::iter(events))
+                            }
+                            None => None,
+                        }
```

Comment:
I think this is just a `map`?
```suggestion
                        context.on_response(&url, &header, &body)
                            .map(|events| {
                                emit!(HttpScrapeEventsReceived {
                                    byte_size: events.size_of(),
                                    count: events.len(),
                                    url: url.to_string()
                                });
                                stream::iter(events)
                            })
```